test/oqs_test_libctx.c

```diff
@@ -0,0 +1,563 @@
+                                         &secdec2, &out2, &outlen2);
+
+    testresult &= oqs_cmp_classical_keys(key1, key2) && (seclen1 == seclen2) &&
+                  (outlen1 == outlen2);
+    if (!testresult)
+        goto err;
+
+    testresult &= oqs_cmp_kem_elems(kemalg_name, secenc1, secenc2, seclen1,
+                                    out1, out2, outlen1);
+
+err:
+    EVP_PKEY_free(key1);
+    EVP_PKEY_free(key2);
+    OPENSSL_free(out1);
+    OPENSSL_free(out2);
+    OPENSSL_free(secenc1);
+    OPENSSL_free(secenc2);
+    OPENSSL_free(secdec1);
+    OPENSSL_free(secdec2);
+    return testresult;
+}
+
+/** \brief Executes the complete comparison of two SIG executions
+ *
+ * \param sigalg_name algorithm name.
+ *
+ * \return 1 if the compare operation is successful, else 0. */
+static int test_oqs_sigs_libctx(const char *sigalg_name) {
+    EVP_PKEY *key1 = NULL, *key2 = NULL;
+    const char msg[] = "The quick brown fox jumps over... you know what";
+    unsigned char *sig1 = NULL, *sig2 = NULL;
+    size_t sig1len, sig2len;
+
+    int testresult = 1;
+
+    if (!alg_is_enabled(sigalg_name)) {
+        printf("Not testing disabled algorithm %s.\n", sigalg_name);
+        return 1;
+    }
+    testresult &= oqs_generate_sig_elems(sigalg_name, msg, sizeof(msg), &key1,
+                                         &sig1, &sig1len) &&
+                  oqs_generate_sig_elems(sigalg_name, msg, sizeof(msg), &key2,
+                                         &sig2, &sig2len);
+
+    if (is_signature_algorithm_hybrid(sigalg_name)) {
+        testresult &= oqs_cmp_classical_keys(key1, key2);
+    } else {
+        testresult &= oqs_cmp_composite_sig_keys(sigalg_name, key1, key2);
+    }
+
+    if (!testresult)
+        goto err;
+
+    testresult &= oqs_cmp_sig_elems(sigalg_name, sig1, sig1len, sig2, sig2len);
+
+err:
+    EVP_PKEY_free(key1);
+    EVP_PKEY_free(key2);
+    OPENSSL_free(sig1);
+    OPENSSL_free(sig2);
+    return testresult;
+}
+
+#define nelem(a) (sizeof(a) / sizeof((a)[0]))
+
+int main(int argc, char *argv[]) {
+    size_t i;
+    int errcnt = 0, test = 0, query_nocache;
+    OSSL_PROVIDER *oqsprov = NULL;
+    const OSSL_ALGORITHM *kemalgs, *sigalgs;
+
+    T((libctx = OSSL_LIB_CTX_new()) != NULL);
+    T(argc == 3);
+    modulename = argv[1];
+    configfile = argv[2];
+
+    oqs_load_det_pseudorandom_generator(libctx);
+    load_oqs_provider(libctx, modulename, configfile);
+
+    oqsprov = OSSL_PROVIDER_load(libctx, modulename);
+
+    kemalgs =
+        OSSL_PROVIDER_query_operation(oqsprov, OSSL_OP_KEM, &query_nocache);
+    if (kemalgs) {
+        for (; kemalgs->algorithm_names != NULL; kemalgs++) {
+            if (!is_kem_algorithm_hybrid(kemalgs->algorithm_names)) {
+                continue;
+            }
+            if (test_oqs_kems_libctx(kemalgs->algorithm_names)) {
+                fprintf(stderr, cGREEN "  KEM test succeeded: %s" cNORM "\n",
```

Comment:
suggest using the label "libctx KEM test" to differentiate from the original KEM test.